### ISSUE TITLE
DetectNet on Custom Dataset

### ISSUE THREAD

**Issue by tamasbalassa** [NVIDIA/DIGITS#1678]
Hello there!

I would kindly ask for some help or suggestions to modify (or even correct/improve) my set up and be able to detect the required object with the help of DIGITS and DetectNet. I would really appreciate your help @lukeyeager, @jbarker-nvidia

So there are the base facts:
- I have same sized (1392x1040 px), grayscale images
- each image contains more objects that we are looking for
- the size of the objects dont change too much (say ~10%)
- the shape of the required object can vary (amoebae-like)
- the images are stack images, which means that they are taken from different z coordinate
- this includes that some of the objects on the image are in focus and some are not
- Im using the setup that is used for car detection at the KITTI database
- I want to detect those objects on the images that are in focus

What I've tried so far / experiments with DetectNet:
### Labeling 
- I was using two labels: dontcare, target
- example row from the labelfile: target 0 0 0 375 319 675 619 0 0 0 0 0 0 0
- the objects that are out of focus are labeled as dontcare
### Bounding Boxes 
- this is a quite hard task to make it perfectly. The bboxes are slightly bigger than the possible biggest objects (Y direction/bottom and up - max +15%; X direction/right and left - max 25%). The reason for this is, that for selecting the correct objects (that are in focus), a framework is used. This framework allows the user to select the center of the object and it will generate a **fix-sized** bounding box around it - which is converted to the proper DIGITS-label format.
###   Image size
- I tried to play a little bit with padding and resizing the images to the KITTI example size (so I dont have to modify anything in the DNN-architecture);
- also changed the size of the inital images (meanwhile creating them), but everything was useless, my mAP stucked at zero. 
- most of the time I got the same / similar results, see:
![screenshot from 2017-06-12 16-30-06](https://user-images.githubusercontent.com/3785204/27038940-b968206c-4f8c-11e7-97da-6eeae5241ee1.png)

Could be the bounding boxes be the problem? Or the wrong size of images? Do I need to modify the architecture to be able to process these images? Is it even possible to detect in-focus objects where out of focus objects are also present or I should not invest any more time in this?

**Comment by ontheway16** [comment]
Hi, try a lower learning rate. ie, 00000.1 instead of  0.00001. 
Also there is no information on parameters you applied. Batch, accumulation, pretained model, num of objects in train and validation dataset, etc. These might help people to eval. your issue. Regards,

**Comment by tamasbalassa** [comment]
Thank You @ontheway16, I will try with a lower lr.

As I mentioned above "Im using the setup that is used for car detection at the KITTI database", which includes the setup from this example: https://github.com/NVIDIA/DIGITS/tree/master/examples/object-detection

Specifically:
- No. objects in train: 1680
- No. objects in val: 150
- Batch size: 2
- Batch accumulation: 5
- Mean: none
- Learning rate: 0.0001
- Solver: Adam
- Pretrained model: bvlc_googlenet
 
Thank You!

**Comment by ontheway16** [comment]
According to sample data line you provided, label box sizes seem within detectnet limits (50-400 pixels for both width & height). But number of samples may stay on low side, especially if objects are visually very different, according to my observations. You may try data augmentation, I believe it may help you a lot. Try adding 90, 180 and 270 deg. rotated versions of current images to the dataset. Try to define a minimum image size and avoid padding and resizing, if possible. There are software available for such data augmentation purposes, you may give it a try.
alpslabel.wordpress.com
Many other useful tools exist, for labeling and augmentation. Please check web.

**Comment by tamasbalassa** [comment]
Thanks again for your suggestions @ontheway16!
I've tried to lower the learning rate to 0.00001 (as You suggested before), but the result was similar to the previous.

The possible weak points that I could imagine are:
- low number of trained samples
- wrongly suited bounding boxes
- wrong size of the input images (?!)
- to make difference between the objects that are in focus and the object that are out of focus is too hard (maybe impossible)

The website that You have linked in your previous comment could hopefully help in the first tree points, however I don't know any possible solutions which can help me in the fourth one. 
First I will try out to fix the first three point and will report the results. Thank You for your time!

**Comment by ontheway16** [comment]
For the last point, my general understanding is, if you are able to visually differentiate these two, then the algorithm can do it also. Therefore I believe other measures will likely to help to the problem.

**Comment by tamasbalassa** [comment]
You were right @ontheway16, the last point was not the problem. What I figured out is that based on the low number of trained samples, I needed many epochs finally to get >0 mAP.

Another thing, when I fixed (resized) the bounding boxes, it made the learner harder to learn the correct objects (before I had a fix sized square bounding boxes). 
Also, "the only thing" about the input images that the architecture requires (but please correct me if Im wrong), that both dimensions should be dividable by 16.

**Comment by dagoPA** [comment]
Did you got any better results? I am facing the same problem
